.travis.yml

```diff
@@ -1,20 +1,48 @@
 language: php
-
 matrix:
-  fast_finish: true
-  include:
-    - php: 7.1
-      env:
-       - COMPOSER_ARG=--prefer-lowest
-    - php: 7.4
-    - php: nightly
-      env:
-       - COMPOSER_ARG=--ignore-platform-reqs
-
+    fast_finish: true
+    include:
+        -
+            php: 7.1
+            env:
+                - COMPOSER_ARG=--prefer-lowest
+        -
+            php: 7.4
+        -
+            php: 7.4
+            env:
+              - PHPUNIT_COVERAGE_TEST=1
+        -
+            php: nightly
+            env:
+                - COMPOSER_ARG=--ignore-platform-reqs
+        -
+            php: 7.4
+            env: PHPCS_TEST=1
+        -
+            php: 7.4
+            env: DUPLICATE_CODE_CHECK=1
+        -
+            php: 7.4
+            env: LINT_CHECK=1
+        -
+            php: 7.4
+            env: PHPSTAN_TEST=1
+        -
+            php: 7.4
+            env: PSALM_TEST=1
 before_script:
-  - composer validate
-  - composer update $COMPOSER_ARG --no-interaction --no-progress --no-suggest --optimize-autoloader --verbose --profile
-
+    - 'composer validate'
+    - 'composer update $COMPOSER_ARG --no-interaction --no-progress --no-suggest --optimize-autoloader --verbose --profile'
+    - 'if [[ $DUPLICATE_CODE_CHECK ]]; then sudo apt remove -y nodejs && curl -sL https://deb.nodesource.com/setup_14.x -o nodesource_setup.sh && sudo bash nodesource_setup.sh && sudo apt install -y build-essential nodejs  && npm install jscpd@3.2.1  ;fi'
 script:
-  - vendor/bin/phpunit
-  - vendor/bin/phpcs src tests
+    - vendor/bin/phpunit
+    - 'if [[ $PHPUNIT_COVERAGE_TEST ]]; then phpdbg -qrr vendor/bin/phpunit tests --coverage-clover=coverage.xml '''' flush=all; fi'
+    - 'if [[ $PHPCS_TEST ]]; then vendor/bin/phpcs --standard=ruleset.xml --extensions=php --tab-width=4 -sp src tests ; fi'
+    - 'if [[ $DUPLICATE_CODE_CHECK ]]; then node_modules/jscpd/bin/jscpd -t 1 src && node_modules/jscpd/bin/jscpd -t 1 tests ; fi'
+    - 'if [[ $LINT_CHECK ]]; then vendor/bin/parallel-lint src/ tests/ ; fi'
```

Comment:
I would have a single config for all the linters. Otherwise it's just a lot more composer time.

Reply:
Interesting idea, although in this case build times are minimal.  I can see the motivation from the perspective of a SilverStripe module though.  I did try ` hirak/prestissimo ` (composer package) which downloads composer packages in parallel, but it made the main SilverStripe Travis test slower :(

Reply:
Sure, for example, the link check is only 59s.

But of that 59s only 0.16s is the actual lint operation the rest is waste as it's setting up the environment. You wouldn't have that execution cost if we combined the linters.

99.7% waste seems bad, even on a fast build ;-)

This means it could also be combined into a single 'composer run-script lint' command which makes it easier to do all the non-test checks at once